src/crimson/osd/main.cc

```diff
@@ -63,8 +64,18 @@ int main(int argc, char* argv[])
       }).then([&conf_file_list] {
         return local_conf().parse_config_files(conf_file_list);
       }).then([&] {
-        return osd.start_single(std::stoi(local_conf()->name.get_id()),
-                                static_cast<uint32_t>(getpid()));
+        int id = std::stoi(local_conf()->name.get_id());
+        auto nonce = static_cast<uint32_t>(getpid());
+        return seastar::when_all_succeed(
+            ceph::net::Messenger::create(entity_name_t::OSD(id), "cluster", nonce, 0),
+            ceph::net::Messenger::create(entity_name_t::OSD(id), "public", nonce, 0),
+            ceph::net::Messenger::create(entity_name_t::OSD(id), "hb_front", nonce, 0),
+            ceph::net::Messenger::create(entity_name_t::OSD(id), "hb_back", nonce, 0)
```

Comment:
> why can't we create messengers where the they are used?

I agree. Actually we can, and we should create messengers where they are used in the 1:1 mapping prototype, so that a compatible-mode messenger can be correctly created, while, be limited within the core where the OSD/Heartbeat/Monc initiated. Please refer to the updated code.

> i'd prefer managing the life cycle of Messenger explicitly by its consumer by calling start() and stop(), and only use ceph::net::Messenger::create() in the unittest and demo.

The purpose of `ceph::net::Messenger::create()` is to allocate `Messenger` in all available cores, allow it to listen and accept from all cores after started, so it won't stop working if reactors > 1. It just make our 1:1 mapping prototype more robust, regardless whether we decide to run OSD instance as shards, or to introduce another reactor to offload some tasks. `start()` and `shutdown()` are still used to let consumer manage the lifecycle of its Messenger. The only difference is we can no longer allocate Messenger inside the constructor of OSD/Heartbeat, because `ceph::net::Messenger::create()` returns a future now.

> as the gracefully shutdown would be a nightmare if we leave it to the create() wrapper.

That's why I decouple `shutdown()` with `stop()`, it allows consumer to end the lifecycle of its Messenger explicitly and gracefully. And the final destruction will be delayed until the engine exits, not with the destruction of the consumer.

Reply:
i see. the life cycle of a shared service is managed by its container, namely, `seastar::sharded<>` and friends. and `seastar::shared<>` does not allow us to update an `seastar::sharded<>` with  a service instance created as a member variable.

what we can is but to promote these member variables to "global" variables even if they are only used by a single class, which does not have to be sharded. but if it is, the `Messenger` should be sharded as well.